src/sinks/datadog/logs/sink.rs

```diff
@@ -137,6 +153,76 @@ pub fn normalize_event(event: &mut Event) {
     }
 }
 
+// Optionally for all other non-reserved fields, nest these under the `message` key. This is the
+// final step in having the event conform to the standard that the logs intake expects when an
+// event originates from an agent. Normalizing the events to the format prepared by the datadog
+// agent resolves any inconsistencies that would be observed when data flows through vector
+// before being ingested by the logs intake. This is because the logs intake interprets the
+// request with slight differences when this header and format are observed.
+pub fn normalize_as_agent_event(event: &mut Event) {
+    let log = event.as_mut_log();
+
+    // Extract the `message` object from the log if it already exists. Then JSON deserialize its
+    // value field. All non reserved fields placed at the root will be moved there.
+    let mut message = if let Some(outer_msg) = log.remove(event_path!(MESSAGE)) {
+        if let Value::Object(current_message) = outer_msg {
+            current_message // already JSON
+        } else {
+            // possible payload is stringified JSON
+            outer_msg
+                .as_bytes()
+                .and_then(|b| serde_json::from_slice::<ObjectMap>(b).ok())
+                .unwrap_or_else(|| {
+                    [(KeyString::from(MESSAGE), outer_msg)]
+                        .into_iter()
+                        .collect()
+                })
+        }
+    } else {
+        ObjectMap::default()
+    };
+
+    // Move all non reserved fields into a new object
+    if let Some(object_map) = log.as_map_mut() {
+        let mut collisions = ObjectMap::default();
+
+        let keys_to_move = object_map
+            .keys()
+            .filter(|key| {
+                DD_RESERVED_SEMANTIC_ATTRS
+                    .iter()
+                    .all(|(_, attr)| *attr != key.as_str())
+            })
+            .cloned()
+            .collect::<Vec<_>>();
+        for key in keys_to_move {
+            if let Some((entry_k, entry_v)) = object_map.remove_entry(key.as_str()) {
+                if let Some(returned_entry_v) = message.insert(entry_k, entry_v) {
+                    collisions.insert(key, returned_entry_v);
+                }
+            }
+        }
```

Comment:
Could this process be done the opposite way:
1. `take` the root object, replacing it with an empty object
2. For each key in the reserved attributes, remove it from the former root into the new root.
3. Insert the remainder into `message` and re-insert `message` into the root.

This eliminates the repeated scans over the reserved attributes and the creation of any temporaries.

Reply:
I just prototyped this, ends up being more code (2 loops one for step 2 and one for step 3) and the temporaries (keys_to_move) I believe must stay. That is because you cannot call `.remove` on a map that you're iterating over and for step 3 (above) we would have to iterate over all remaining keys, calling `.remove` on the same map. The main drawback being repeated scans over the reserved attrs list.

Reply:
What I was thinking was something like this, which seems different than what you described:
```rust
let old_root = std::mem::take(&object_map);
for key in DD_RESERVED_SEMANTIC_ATTRS {
    if let Some((key, value)) = old_root.remove_entry(key) {
        object_map.insert(key, value); // will never be `Some`
    }
}
for (key, value) in old_root {
    if let Some(returned_entry_v) = message.insert(key, value) {
        collisions.insert(key, returned_entry_v);
    }
}
object_map.insert(MESSAGE, message);
````